### PR TITLE
Add url argument for conda channel mirrors

### PR DESCRIPTION
**Pull request by quartox**
Would like to talk about `url` argument name (I think there is probably a better name) and what testing needs to be done.

I realized the use case is more complicated than I first thought. Our mirror follows the pattern `http://internal.url.com/conda-forge-remote` so this PR allows `python app.py --url https://internal.url.com` and then appends `channel` and `arch` or `python app.py --url https://internal.url.com/{channel}-remote/` or `python app.py --url https://internal.url.com/{channel}-remote/{arch}` in case the channel and/or arch are embedded in the middle of the url.

I have manually proven that this will work on linux using the following:

`$ python --url https://internal.url.com/{channel}-remote/`

in another terminal

`$ conda create --name metachannel-test python=3.7 pandas -c http://127.0.0.1:20124/conda-forge/pandas`

### CODE REVIEW

**Comment by scopatz** [comment]
Maybe call it base_url?

**Comment by scopatz** [comment]
Or if you just wanted `internal.url.com`, and not `http://internal.url.com`, you could call it `tld` or `domain`

File: app.py [hunk 1]
```diff
@@ -129,8 +129,11 @@ def in_container():
     parser.add_argument("--host", default="127.0.0.1")
     parser.add_argument("--port", default=20124, type=int)
     parser.add_argument("--reload", action="store_true")
+    parser.add_argument("--url", default="https://conda.anaconda.org/")
     args = parser.parse_args()
 
+    URL = args.url.rstrip('/')
```
Comment (scopatz) [inline app.py]: variable names should be `lower_case_with_underscores`

File: graph.py [hunk 2]
```diff
@@ -89,12 +88,17 @@ class RawRepoData:
     _cache = TTLCache(100, ttl=_ttl)
     _last_expiry = time.monotonic()
 
-    def __init__(self, channel: str, arch: str = "linux-64", ttl=600):
+    def __init__(self, url: str, channel: str, arch: str = "linux-64", ttl=600):
```
Comment (scopatz) [inline graph.py]: `url` (or whatever it ends up being called), should be a keyword argument with `"https://conda.anaconda.org/"` as the default.

File: app.py (hunk 1 above)
Comment (quartox) [reply app.py]: I believe this one counts as a global module variable and I use it as a global in line 31. I did not see a way to explicitly pass it `fetch_artifact_graph` (though I may be mistaken) and I felt it was less confusing to use global variable convention. I am still happy to lowercase it if that is preferred.

Comment (scopatz) [reply app.py]: Yeah, it is inconsistent with parser, args, etc as it stands.  Really everything here should be in a main() function, but I'm not asking you to rearrange here :wink:

**Comment by quartox** [comment]
I like `base_url` since it can be a url pattern in the case where the user needs to include `{channel}`. In my use case I also need to pass `http://internal.company.com/artifactory/{channel}-remote` so I think `domain` or `tld` might be more confusing.

**Comment by scopatz** [comment]
Sounds good to me! Let us know when this is ready fro another review!

**Comment by quartox** [comment]
Ready for another review. I ended up putting the default in each function/class that gets passed the `base_url` but that may have been a misunderstanding of your comment. I never know if I should have the default in each function or only certain ones, happy to change it to whatever you would prefer.

**Comment by scopatz** [comment]
default in each seems preferable.

**Comment by scopatz** [comment]
This LGTM  @mariusvniekerk @CJ-Wright any thoughts?

File: graph.py [hunk 3]
```diff
@@ -89,12 +88,17 @@ class RawRepoData:
...
         self.ttl = ttl
         # normal seetings
         logger.info(f"RETRIEVING: {channel}, {arch}")
-        url_prefix = f"https://conda.anaconda.org/{channel}/{arch}"
+        if '{channel}' in base_url and '{arch}' in base_url:
+            url_prefix = base_url.format(channel=channel, arch=arch)
+        elif '{channel}' in base_url:
+            url_prefix = base_url.format(channel=channel) + f"/{arch}"
+        else:
+            url_prefix = f"{base_url}/{channel}/{arch}"
```
Comment (mariusvniekerk) [inline graph.py]: double slashes here in the `conda.anaconda.org/` case

Comment (scopatz) [reply graph.py]: Agreed, there used to be an rstip here

**Pull request merged**